core/typedefs.h

```diff
@@ -440,3 +440,16 @@ inline constexpr bool is_zero_constructible_v = is_zero_constructible<T>::value;
 #define GODOT_MSVC_WARNING_POP
 #define GODOT_MSVC_WARNING_PUSH_AND_IGNORE(m_warning)
 #endif
+
+template <typename T, typename = void>
+struct is_incomplete_type : std::true_type {};
+
+template <typename T>
+struct is_incomplete_type<T, std::void_t<decltype(sizeof(T))>> : std::false_type {};
+
+template <typename T>
+constexpr bool is_incomplete_type_v = is_incomplete_type<T>::value;
+
+#define STATIC_ASSERT_INCOMPLETE_TYPE(m_type) \
+	class m_type;                             \
+	static_assert(is_incomplete_type_v<m_type>, #m_type " was defined when it should not be. Please check include hierarchy of " __FILE__);
```

Comment:
Ah I can see a problem here with SCU builds. It will depend on the build parameters whether in some cases `display_types.h` might get included before the cpp I think?

Reply:
I think the checks are absolutely necessary, otherwise we'll be hunting include regressions forever.
An SCU define would totally work 👍

Reply:
Ok remind me tommorrow on RocketChat, I'll take a look into it if you can't work out how to (I'm not super familiar with the build files and `#defines` but I'm sure we can work it out).

Reply:
I'm actually away over the weekend, so feel free to take it 👍

Reply:
I think I got it working